test/test_calendar.py

```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+        [
+            (
+                datetime.datetime(year=4, month=1, day=1),
+                datetime.datetime(year=5, month=1, day=1),
+                GREGORIAN,
+                366 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=4, month=1, day=1),
+                datetime.datetime(year=5, month=1, day=1),
+                NOLEAP,
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=300, month=1, day=1),
+                datetime.datetime(year=301, month=1, day=1),
+                GREGORIAN,
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=400, month=1, day=1),
+                datetime.datetime(year=401, month=12, day=31),
+                GREGORIAN,
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=12, month=7, day=22),
+                datetime.datetime(year=23, month=3, day=15),
+                GREGORIAN,
+                (10 * 365 + 238) * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=1, month=1, day=1),
+                datetime.datetime(year=1001, month=1, day=1),
+                GREGORIAN,
+                (1000 * 365 + 242) * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=1, month=1, day=1),
+                datetime.datetime(year=9999, month=1, day=1),
+                GREGORIAN,
+                (9998 * 365 + 2424) * SEC_PER_DAY
+            )
+        ]
+)
+def test_seconds_between_dates(start_date, end_date, caltype_int, expected):
+    assert seconds_between_dates(start_date, end_date, caltype_int) == expected
+
+
+@pytest.mark.parametrize(
+        "date_int, expected",
+        [
+            (10101, datetime.date(1, 1, 1)),
+            (100321, datetime.date(10, 3, 21)),
+            (38211130, datetime.date(3821, 11, 30))
+        ]
+)
+def test_int_to_date(date_int, expected):
+    """
+    Check that integers typically read in from namelists
+    are correctly converted to datetime.date objects.
+    """
+    converted_date = int_to_date(date_int)
+    assert converted_date == expected
+
+
+@pytest.mark.parametrize(
+        "start_date, years, months, days, seconds, caltype, expected",
+        [
+            (datetime.date(101, 1, 1), 1, 0, 0, 0, GREGORIAN, 365*SEC_PER_DAY),
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, GREGORIAN, 29*SEC_PER_DAY),
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, NOLEAP, 28*SEC_PER_DAY),
+            (datetime.date(153, 8, 21), 2, 7, 510, 5321,
+             GREGORIAN, 1453*SEC_PER_DAY + 5321),
+            # Extreme cases, unlikely to ever happen
+            (datetime.date(1, 1, 1), 9998, 0, 0, 0,
+             NOLEAP, 9998 * 365 * SEC_PER_DAY),
+            (datetime.date(1, 1, 1), 9998, 0, 0, 0,
+             GREGORIAN, (9998 * 365 + 2424) * SEC_PER_DAY)
+        ]
+)
```

Comment:
I think this would be clearer if you just have the limits in all permutations:

For start_date: the minimum (e.g. datetime.date(1,1,1)) and the maximum (datetime.date(9999,12,31)) 
For run time: the minimum (1 second) and the maximums, (9999,0,0,0) and (9998,12,31,86400)
For calendar: GREGORIAN and NOLEAP

Its about 12 cases.

Reply:
Good idea to test around the edges here. I think it's worth retaining the cases like `(datetime.date(101, 1, 1), 1, 0, 0, 0, GREGORIAN, 365*SEC_PER_DAY)`, as these match what people will generally use for ESM1.5, e.g. 1 year runtimes, starting at the beginning of the year. 

In terms of the max cases, start dates like `datetime.date(9999,12,31)` cause issues, as the `runtime_from_date` function internally creates an end date of `start date + runtime`, which then goes outside the limits of `datetime.date`. I've added in a few different limiting cases though – let me know if you think they cover what we need.

Reply:
Im just going to go ahead and assume no paleo folk do a climate run longer than 9999 years !

Reply:
This is much clearer, thankyou

Reply:
Turns out there was already an issue for the 9999 years thing ! https://github.com/payu-org/payu/issues/305